File: Tests/SharedCoreTrafficShadowComparatorTests.swift
import XCTest

@testable import MihomoMeter

final class SharedCoreTrafficShadowComparatorTests: XCTestCase {
  func testComparatorMatchesAllProductionFormats() {
    let scaleTraffic: (UInt64) throws -> SharedTrafficScale = { _ in
      SharedTrafficScale(
        value: 1.5,
        unit: .kilobytes,
        decimalPlaces: 2
      )
    }

    XCTAssertEqual(
      SharedCoreTrafficShadowComparator.compare(
        bytes: 1_500,
        nativeText: "1.50 KB",
        format: .byteCount,
        scaleTraffic: scaleTraffic
      ),
      .matched
    )
    XCTAssertEqual(
      SharedCoreTrafficShadowComparator.compare(
        bytes: 1_500,
        nativeText: "1.50 KB/s",
        format: .rate,
        scaleTraffic: scaleTraffic
      ),
      .matched
    )
    XCTAssertEqual(
      SharedCoreTrafficShadowComparator.compare(
        bytes: 1_500,
        nativeText: "1.50K/s",
        format: .compactRate,
        scaleTraffic: scaleTraffic
      ),
      .matched
    )
  }

  func testComparatorReportsMismatchWithoutChangingNativeText() {
    let status = SharedCoreTrafficShadowComparator.compare(
      bytes: 1_500,
      nativeText: "原生输出",
      format: .byteCount,
      scaleTraffic: { _ in
        SharedTrafficScale(
          value: 1.5,
          unit: .kilobytes,
          decimalPlaces: 2
        )
      }
    )

    XCTAssertEqual(status, .mismatch)
  }

  func testComparatorMapsAdapterFailuresToCoarseStatuses() {
    XCTAssertEqual(
      compareThrowing(.unsupportedABIVersion(2)),
      .abiMismatch
    )
    XCTAssertEqual(
      compareThrowing(.nativeCallFailed(-1)),
      .nativeCallFailed
    )
    XCTAssertEqual(
      compareThrowing(.unsupportedTrafficUnit(99)),
      .unsupportedTrafficUnit
    )
  }

  func testComparatorRejectsUnsupportedDecimalPlaces() {
    let status = SharedCoreTrafficShadowComparator.compare(
      bytes: 1_500,
      nativeText: "1.50 KB",
      format: .byteCount,
      scaleTraffic: { _ in
        SharedTrafficScale(
          value: 1.5,
          unit: .kilobytes,
          decimalPlaces: 3
        )
      }
    )

    XCTAssertEqual(status, .unexpectedResult)
  }

  func testObservationGateReportsEachFormatAndStatusOnce() {
    var gate = SharedCoreTrafficShadowObservationGate()
    let byteCountMatched = SharedCoreTrafficShadowObservation(
      format: .byteCount,
      status: .matched
    )

    XCTAssertTrue(gate.shouldReport(byteCountMatched))
    XCTAssertFalse(gate.shouldReport(byteCountMatched))
    XCTAssertTrue(
      gate.shouldReport(
        SharedCoreTrafficShadowObservation(
          format: .byteCount,
          status: .mismatch
        )
      )
    )
    XCTAssertTrue(
      gate.shouldReport(
        SharedCoreTrafficShadowObservation(
          format: .rate,
          status: .matched
        )
      )
    )

    gate.reset()
    XCTAssertTrue(gate.shouldReport(byteCountMatched))
  }

  private func compareThrowing(
    _ error: SharedCoreAdapterError
  ) -> SharedCoreTrafficShadowStatus {
    SharedCoreTrafficShadowComparator.compare(
      bytes: 1_500,
      nativeText: "1.50 KB",
      format: .byteCount,
      scaleTraffic: { _ in throw error }
    )
  }
}

File: platform/windows/MihomoMeter.Windows.Tests/SharedCoreTrafficShadowComparatorTests.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using MihomoMeter.Windows.Core.Application;

namespace MihomoMeter.Windows.Tests;

[TestClass]
public sealed class SharedCoreTrafficShadowComparatorTests
{
    [TestMethod]
    [DataRow(SharedCoreTrafficFormat.ByteCount, "1.50 KB")]
    [DataRow(SharedCoreTrafficFormat.Rate, "1.50 KB/s")]
    [DataRow(SharedCoreTrafficFormat.CompactRate, "1.50K/s")]
    public void ComparatorMatchesEveryProductionFormat(
        SharedCoreTrafficFormat format,
        string nativeText)
    {
        var status = SharedCoreTrafficShadowComparator.Compare(
            1_500,
            nativeText,
            format,
            () => 1,
            _ => new SharedTrafficScale(
                1.5,
                SharedTrafficUnit.Kilobytes,
                2));

        Assert.AreEqual(SharedCoreTrafficShadowStatus.Matched, status);
    }

    [TestMethod]
    public void ComparatorStopsBeforeScalingForAbiMismatch()
    {
        var didScaleTraffic = false;

        var status = SharedCoreTrafficShadowComparator.Compare(
            1_500,
            "1.50 KB",
            SharedCoreTrafficFormat.ByteCount,
            () => 2,
            _ =>
            {
                didScaleTraffic = true;
                return new SharedTrafficScale(
                    1.5,
                    SharedTrafficUnit.Kilobytes,
                    2);
            });

        Assert.AreEqual(SharedCoreTrafficShadowStatus.AbiMismatch, status);
        Assert.IsFalse(didScaleTraffic);
    }

    [TestMethod]
    public void ComparatorMapsNativeCallFailureWithoutThrowing()
    {
        var status = SharedCoreTrafficShadowComparator.Compare(
            1_500,
            "1.50 KB",
            SharedCoreTrafficFormat.ByteCount,
            () => throw new DllNotFoundException("synthetic"),
            _ => throw new InvalidOperationException("不应调用。"));

        Assert.AreEqual(SharedCoreTrafficShadowStatus.NativeCallFailed, status);
    }

    [TestMethod]
    public void ComparatorReportsMismatchWithoutReturningSharedText()
    {
        var status = SharedCoreTrafficShadowComparator.Compare(
            1_500,
            "原生输出",
            SharedCoreTrafficFormat.ByteCount,
            () => 1,
            _ => new SharedTrafficScale(
                1.5,
                SharedTrafficUnit.Kilobytes,
                2));

        Assert.AreEqual(SharedCoreTrafficShadowStatus.Mismatch, status);
    }

    [TestMethod]
    public void ComparatorRejectsUnsupportedDecimalPlaces()
    {
        var status = SharedCoreTrafficShadowComparator.Compare(
            1_500,
            "1.50 KB",
            SharedCoreTrafficFormat.ByteCount,
            () => 1,
            _ => new SharedTrafficScale(
                1.5,
                SharedTrafficUnit.Kilobytes,
                3));

        Assert.AreEqual(SharedCoreTrafficShadowStatus.UnexpectedResult, status);
    }

    [TestMethod]
    public void ShadowReturnsNativeTextWhenDiagnosticReporterFails()
    {
        SharedCoreTrafficShadow.ConfigureReporter(_ => throw new InvalidOperationException(
            "synthetic"));
        try
        {
            Assert.AreEqual(
                "原生输出",
                SharedCoreTrafficShadow.Observe(
                    1_500,
                    "原生输出",
                    SharedCoreTrafficFormat.ByteCount));
        }
        finally
        {
            SharedCoreTrafficShadow.ConfigureReporter(null);
        }
    }
}
